lib/utils/replace.go

```diff
@@ -74,15 +90,28 @@ func RegexpWithConfig(expression string, config RegexpConfig) (*regexp.Regexp, e
 	if err != nil {
 		return nil, trace.BadParameter("%s", err)
 	}
+
+	regexpCache.Add(key, expr)
 	return expr, nil
 }
 
-// ReplaceRegexp replaces string in a given regexp.
+// RegexpWithConfig compiles a regular expression given some configuration.
+// There are several important differences with standard lib (see ReplaceRegexp).
+func RegexpWithConfig(expression string, config RegexpConfig) (*regexp.Regexp, error) {
+	expr, err := replaceRegexCached(expression, config)
+	return expr, trace.Wrap(err)
+}
+
+// ReplaceRegexpWith replaces string in a given regexp.
 func ReplaceRegexpWith(expr *regexp.Regexp, replaceWith string, input string) (string, error) {
 	// if there is no match, return NotFound error
 	index := expr.FindStringIndex(input)
 	if index == nil {
-		return "", trace.NotFound("no match found")
+		// The returned error is intentionally not wrapped to avoid
+		// capturing stack traces. This method is used by authorization
+		// logic and the additional overhead of strack trace capturing
+		// is a performance bottleneck.
+		return "", ErrReplaceRegexNotFound
 	}
 	return expr.ReplaceAllString(input, replaceWith), nil
```

Comment:
Is this applying the regexp twice on the same input? Is there a way to use the regexp API to avoid that?